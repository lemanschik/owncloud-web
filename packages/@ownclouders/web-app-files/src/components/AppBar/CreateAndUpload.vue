<template>
  <div v-if="showActions" class="create-and-upload-actions oc-flex-inline oc-width-1-1">
    <template v-if="createFileActionsAvailable">
      <span v-oc-tooltip="newButtonTooltip">
        <oc-button
          id="new-file-menu-btn"
          key="new-file-menu-btn-enabled"
          v-oc-tooltip="hideButtonLabels ? $gettext('New') : ''"
          :aria-label="newButtonAriaLabel"
          appearance="filled"
          variation="primary"
          :disabled="uploadOrFileCreationBlocked"
        >
          <oc-icon name="add" />
          <span v-if="!hideButtonLabels" v-text="$gettext('New')" />
        </oc-button>
      </span>
      <oc-drop
        drop-id="new-file-menu-drop"
        toggle="#new-file-menu-btn"
        mode="click"
        close-on-click
        class="oc-width-auto"
        padding-size="small"
      >
        <oc-list id="create-list">
          <li class="create-list-folder oc-menu-item-hover">
            <oc-button id="new-folder-btn" appearance="raw" @click="createNewFolderAction">
              <oc-resource-icon :resource="folderIconResource" size="medium" />
              <span v-text="$gettext('Folder')" />
            </oc-button>
          </li>
          <li
            v-for="(fileAction, key) in createNewFileActions"
            :key="`file-creation-item-${key}`"
            class="create-list-file oc-menu-item-hover"
          >
            <oc-button
              appearance="raw"
              :class="['new-file-btn-' + fileAction.ext]"
              @click="fileAction.handler"
            >
              <oc-resource-icon :resource="getIconResource(fileAction)" size="medium" />
              <span>{{ fileAction.label() }}</span>
            </oc-button>
          </li>
          <template v-if="mimetypesAllowedForCreation">
            <li
              v-for="(mimeTypeAction, key) in createNewFileMimeTypeActions"
              :key="`file-creation-item-external-${key}`"
              class="create-list-file oc-menu-item-hover"
            >
              <oc-button appearance="raw" @click="mimeTypeAction.handler">
                <oc-resource-icon :resource="getIconResource(mimeTypeAction)" size="medium" />
                <span v-text="$gettext(mimeTypeAction.label())" />
              </oc-button>
            </li>
          </template>
        </oc-list>
      </oc-drop>
    </template>
    <template v-else>
      <span v-oc-tooltip="newButtonTooltip">
        <oc-button
          id="new-folder-btn"
          v-oc-tooltip="hideButtonLabels ? $gettext('New Folder') : ''"
          appearance="filled"
          variation="primary"
          :aria-label="newButtonAriaLabel"
          :disabled="uploadOrFileCreationBlocked"
          @click="createNewFolderAction"
        >
          <oc-icon name="resource-type-folder" />
          <span v-if="!hideButtonLabels" v-text="$gettext('New Folder')" />
        </oc-button>
      </span>
    </template>
    <span v-oc-tooltip="uploadButtonTooltip">
      <oc-button
        id="upload-menu-btn"
        key="upload-menu-btn-enabled"
        v-oc-tooltip="hideButtonLabels ? $gettext('Upload') : ''"
        :aria-label="uploadButtonAriaLabel"
        :disabled="uploadOrFileCreationBlocked"
      >
        <oc-icon name="upload" fill-type="line" />
        <span v-if="!hideButtonLabels" v-text="$gettext('Upload')" />
      </oc-button>
    </span>
    <oc-drop
      drop-id="upload-menu-drop"
      toggle="#upload-menu-btn"
      mode="click"
      class="oc-width-auto"
      close-on-click
      padding-size="small"
    >
      <oc-list id="upload-list">
        <li class="oc-menu-item-hover">
          <resource-upload ref="folder-upload" btn-class="oc-width-1-1" />
        </li>
        <li class="oc-menu-item-hover">
          <resource-upload ref="file-upload" btn-class="oc-width-1-1" :is-folder="true" />
        </li>
      </oc-list>
    </oc-drop>
    <div
      v-if="showPasteHereButton"
      id="clipboard-btns"
      v-oc-tooltip="
        uploadOrFileCreationBlocked ? $gettext('You have no permission to paste files here!') : ''
      "
      class="oc-button-group"
      :class="{ disabled: uploadOrFileCreationBlocked }"
    >
      <oc-button
        :disabled="uploadOrFileCreationBlocked"
        class="paste-files-btn"
        @click="pasteFilesHere"
      >
        <oc-icon fill-type="line" name="clipboard" />
        <span v-translate>Paste here</span>
      </oc-button>
      <oc-button
        :disabled="uploadOrFileCreationBlocked"
        class="clear-clipboard-btn"
        @click="clearClipboardFiles"
      >
        <oc-icon fill-type="line" name="close" />
      </oc-button>
    </div>
  </div>
</template>

<script lang="ts">
import { mapActions, mapGetters, mapMutations } from 'vuex'

import { useFileActions } from '../../composables/actions/files/useFileActions'
import { isLocationPublicActive, isLocationSpacesActive } from '../../router'
import {
  useActiveLocation,
  useFileActionsCreateNewFile,
  useFileActionsCreateNewFolder
} from '../../composables'

import {
  useRequest,
  useCapabilityShareJailEnabled,
  useCapabilitySpacesEnabled,
  useStore,
  useUserContext,
  useClientService
} from 'web-pkg/src/composables'

import ResourceUpload from './Upload/ResourceUpload.vue'
import {
  computed,
  defineComponent,
  getCurrentInstance,
  onMounted,
  onBeforeUnmount,
  PropType,
  unref,
  watch
} from 'vue'
import { useUpload } from 'web-runtime/src/composables/upload'
import { useUploadHelpers } from '../../composables/upload'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { Resource, SpaceResource } from 'web-client/src/helpers'
import { ResourcesUpload } from '../../helpers/resource'
import { useService } from 'web-pkg/src/composables/service'
import { UppyService } from 'web-runtime/src/services/uppyService'

export default defineComponent({
  components: {
    ResourceUpload
  },
  props: {
    space: {
      type: Object as PropType<SpaceResource>,
      required: true
    },
    item: {
      type: String,
      required: false,
      default: null
    },
    limitedScreenSpace: {
      type: Boolean,
      default: false,
      required: false
    },
    itemId: {
      type: [String, Number],
      required: false,
      default: null
    }
  },
  setup(props) {
    const instance = getCurrentInstance().proxy as any
    const uppyService = useService<UppyService>('$uppyService')
    const clientService = useClientService()
    const store = useStore()

    let filesSelectedSub
    let uploadCompletedSub

    const { actions: createNewFolder } = useFileActionsCreateNewFolder({
      store,
      space: props.space
    })
    const createNewFolderAction = computed(() => unref(createNewFolder)[0].handler)

    const newFileHandlers = computed(() => store.getters.newFileHandlers)

    const { actions: createNewFileActions } = useFileActionsCreateNewFile({
      store,
      space: props.space,
      newFileHandlers: newFileHandlers
    })

    const mimetypesAllowedForCreation = computed(() => {
      const mimeTypes = store.getters['External/mimeTypes']
      if (!mimeTypes) {
        return []
      }
      return mimeTypes.filter((mimetype) => mimetype.allow_creation) || []
    })

    const { actions: createNewFileMimeTypeActions } = useFileActionsCreateNewFile({
      store,
      space: props.space,
      mimetypesAllowedForCreation: mimetypesAllowedForCreation
    })

    const currentFolder = computed(() => {
      return store.getters['Files/currentFolder']
    })
    const canUpload = computed(() => {
      return unref(currentFolder)?.canUpload({ user: store.getters.user })
    })

    onMounted(() => {
      filesSelectedSub = uppyService.subscribe('filesSelected', instance.onFilesSelected)
      uploadCompletedSub = uppyService.subscribe('uploadCompleted', instance.onUploadComplete)
    })

    onBeforeUnmount(() => {
      uppyService.unsubscribe('filesSelected', filesSelectedSub)
      uppyService.unsubscribe('uploadCompleted', uploadCompletedSub)
      uppyService.removeDropTarget()
    })

    watch(
      canUpload,
      () => {
        if (unref(canUpload)) {
          uppyService.useDropTarget({ targetSelector: '#files-view', uppyService })
        } else {
          uppyService.removeDropTarget()
        }
      },
      { immediate: true }
    )

    return {
      ...useFileActions({ store }),
      ...useUpload({
        uppyService
      }),
      ...useUploadHelpers({
        uppyService,
        space: computed(() => props.space),
        currentFolder: computed(() => props.item),
        currentFolderId: computed(() => props.itemId)
      }),
      ...useRequest(),
      clientService,
      isPublicLocation: useActiveLocation(isLocationPublicActive, 'files-public-link'),
      isSpacesGenericLocation: useActiveLocation(isLocationSpacesActive, 'files-spaces-generic'),
      hasShareJail: useCapabilityShareJailEnabled(),
      hasSpaces: useCapabilitySpacesEnabled(),
      isUserContext: useUserContext({ store }),
      canUpload,
      currentFolder,
      createNewFileActions,
      createNewFileMimeTypeActions,
      createNewFolder,
      mimetypesAllowedForCreation,
      createNewFolderAction
    }
  },
  data: () => ({
    newFileAction: null
  }),
  computed: {
    ...mapGetters(['capabilities', 'configuration', 'newFileHandlers', 'user']),
    ...mapGetters('Files', ['ancestorMetaData', 'files', 'selectedFiles', 'clipboardResources']),
    ...mapGetters('runtime/spaces', ['spaces']),

    showPasteHereButton() {
      return this.clipboardResources && this.clipboardResources.length !== 0
    },
    hideButtonLabels() {
      return this.limitedScreenSpace && this.showPasteHereButton
    },

    showActions() {
      return !(this.uploadOrFileCreationBlocked && this.isPublicLocation)
    },

    createFileActionsAvailable() {
      return this.newFileHandlers.length > 0 || this.mimetypesAllowedForCreation.length > 0
    },
    newButtonTooltip() {
      if (!this.canUpload) {
        return this.$gettext('You have no permission to create new files!')
      }
      return null
    },
    newButtonAriaLabel() {
      const tooltip = this.newButtonTooltip
      if (tooltip) {
        return tooltip
      }
      if (!this.createFileActionsAvailable) {
        return this.$gettext('Create a new folder')
      }
      return this.$gettext('Create new files or folders')
    },
    uploadButtonTooltip() {
      if (!this.canUpload) {
        return this.$gettext('You have no permission to upload!')
      }
      return null
    },
    uploadButtonAriaLabel() {
      const tooltip = this.uploadButtonTooltip
      if (tooltip) {
        return tooltip
      }
      return this.$gettext('Upload files or folders')
    },

    uploadOrFileCreationBlocked() {
      return !this.canUpload
    },

    loadIndicatorsForNewFile() {
      return this.isSpacesGenericLocation && this.space.driveType !== 'share'
    },

    folderIconResource() {
      return { isFolder: true, extension: '' } as Resource
    }
  },
  methods: {
    ...mapActions('Files', ['clearClipboardFiles', 'pasteSelectedFiles']),
    ...mapActions(['showMessage', 'createModal', 'setModalInputErrorMessage', 'hideModal']),
    ...mapMutations('Files', ['UPSERT_RESOURCE']),
    ...mapMutations('runtime/spaces', ['UPDATE_SPACE_FIELD']),
    ...mapMutations(['SET_QUOTA']),

    pasteFilesHere() {
      this.pasteSelectedFiles({
        targetSpace: this.space,
        clientService: this.clientService,
        loadingService: this.$loadingService,
        createModal: this.createModal,
        hideModal: this.hideModal,
        showMessage: this.showMessage,
        $gettext: this.$gettext,
        $gettextInterpolate: this.$gettextInterpolate,
        $ngettext: this.$ngettext
      }).then(() => {
        ;(document.activeElement as HTMLElement).blur()
      })
    },

    async onUploadComplete(result) {
      if (result.successful) {
        const file = result.successful[0]

        if (!file) {
          return
        }

        const { spaceId, currentFolder, currentFolderId } = file.meta
        if (!['public', 'share'].includes(file.meta.driveType)) {
          if (this.hasSpaces) {
            const client = this.clientService.graphAuthenticated
            const driveResponse = await client.drives.getDrive(spaceId)
            this.UPDATE_SPACE_FIELD({
              id: driveResponse.data.id,
              field: 'spaceQuota',
              value: driveResponse.data.quota
            })
          } else {
            const user = await this.$client.users.getUser(this.user.id)
            this.SET_QUOTA(user.quota)
          }
        }

        const sameFolder = this.itemId
          ? currentFolderId === this.itemId
          : currentFolder === this.item
        const fileIsInCurrentPath = spaceId === this.space.id && sameFolder
        if (fileIsInCurrentPath) {
          eventBus.publish('app.files.list.load')
        }
      }
    },

    onFilesSelected(filesToUpload: File[]) {
      const uploader = new ResourcesUpload(
        filesToUpload,
        this.files,
        this.inputFilesToUppyFiles,
        this.$uppyService,
        this.space,
        this.item,
        this.itemId,
        this.spaces,
        this.hasSpaces,
        this.createDirectoryTree,
        this.createModal,
        this.hideModal,
        this.showMessage,
        this.$gettext,
        this.$ngettext,
        this.$gettextInterpolate
      )
      uploader.perform()
    },

    getIconResource(fileHandler) {
      return { type: 'file', extension: fileHandler.ext } as Resource
    }
  }
})
</script>
<style lang="scss">
#create-list {
  li {
    border: 1px solid transparent;

    button {
      gap: 10px;
      justify-content: left;
      width: 100%;
    }
  }

  .create-list-folder {
    border-bottom: 1px solid var(--oc-color-border);
  }

  .create-list-folder button {
    margin-bottom: 8px;
  }

  .create-list-file:nth-child(2) button {
    margin-top: 6px;
  }
}

#upload-list,
#new-file-menu-drop {
  min-width: 230px;
}

#create-list,
#upload-list,
#new-file-menu-drop {
  .oc-icon-m svg {
    height: 100% !important;
  }
}

#clipboard-btns {
  flex-flow: inherit;

  :nth-child(1) {
    border-right: 0px !important;
    white-space: nowrap;
  }

  :nth-child(2) {
    border-left: 0px !important;
  }
}

#clipboard-btns.disabled {
  opacity: 0.6;

  button {
    opacity: 1;
  }
}

.create-and-upload-actions {
  gap: var(--oc-space-small);
}
</style>
