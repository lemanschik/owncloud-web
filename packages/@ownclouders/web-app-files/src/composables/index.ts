export * from './actions'
export * from './archiverService'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './selection'
export * from './shares'
export * from './spaces'
export * from './viewMode'
