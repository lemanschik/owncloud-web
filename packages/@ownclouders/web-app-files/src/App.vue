<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div v-if="dragareaEnabled" class="dragarea" />
    <router-view tabindex="0" class="files-wrapper oc-width-expand oc-height-1-1 oc-flex-wrap" />
  </main>
</template>
<script lang="ts">
import { defineComponent, onBeforeUnmount, watch, ref } from 'vue'
import { useRoute, useStore } from 'web-pkg/src/composables'
import { eventBus } from 'web-pkg/src/services/eventBus'

export default defineComponent({
  setup() {
    const store = useStore<any>()
    const dragareaEnabled = ref(false)
    watch(useRoute(), () => {
      store.dispatch('Files/resetFileSelection')
    })

    const hideDropzone = () => {
      dragareaEnabled.value = false
    }
    const onDragOver = (event) => {
      dragareaEnabled.value = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }

    const dragOver = eventBus.subscribe('drag-over', onDragOver)
    const dragOut = eventBus.subscribe('drag-out', hideDropzone)
    const drop = eventBus.subscribe('drop', hideDropzone)

    onBeforeUnmount(() => {
      eventBus.unsubscribe('drag-over', dragOver)
      eventBus.unsubscribe('drag-out', dragOut)
      eventBus.unsubscribe('drop', drop)
    })
    return { dragareaEnabled }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-color-swatch-primary-muted);
}

#files {
  position: relative;
}
</style>
