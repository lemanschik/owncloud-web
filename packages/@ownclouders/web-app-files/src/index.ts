import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceDriveResolver from './views/spaces/DriveResolver.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import TrashOverview from './views/trash/Overview.vue'
import translations from '../l10n/translations.json'
import quickActions from './quickActions'
import store from './store'
import { SDKSearch } from './search'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { Registry, ArchiverService } from './services'
import fileSideBars from './fileSideBars'
import { buildRoutes } from './router'
import get from 'lodash-es/get'
import { AppNavigationItem, AppReadyHookArgs } from 'web-pkg/src/apps'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'
import { isPersonalSpaceResource } from 'web-client/src/helpers'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
const navItems = (context): AppNavigationItem[] => {
  return [
    {
      name(capabilities) {
        return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
      },
      icon: appInfo.icon,
      route: {
        path: `/${appInfo.id}/spaces/personal`
      },
      enabled(capabilities) {
        if (!capabilities.spaces?.enabled) {
          return true
        }
        return !!context?.$store?.getters['runtime/spaces/spaces'].find((drive) =>
          isPersonalSpaceResource(drive)
        )
      }
    },
    {
      name: $gettext('Favorites'),
      icon: 'star',
      route: {
        path: `/${appInfo.id}/favorites`
      },
      enabled(capabilities) {
        return capabilities.files?.favorites
      }
    },
    {
      name: $gettext('Shares'),
      icon: 'share-forward',
      route: {
        path: `/${appInfo.id}/shares`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/share` }],
      enabled(capabilities) {
        return capabilities.files_sharing?.api_enabled !== false
      }
    },
    {
      name: $gettext('Spaces'),
      icon: 'layout-grid',
      route: {
        path: `/${appInfo.id}/spaces/projects`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/project` }],
      enabled(capabilities) {
        return capabilities.spaces?.projects
      }
    },
    {
      name: $gettext('Deleted files'),
      icon: 'delete-bin-5',
      route: {
        path: `/${appInfo.id}/trash/overview`
      },
      activeFor: [{ path: `/${appInfo.id}/trash` }],
      enabled(capabilities) {
        return capabilities.dav?.trashbin === '1.0'
      }
    }
  ]
}

export default {
  appInfo,
  store,
  routes: buildRoutes({
    App,
    Favorites,
    FilesDrop,
    SearchResults,
    Shares: {
      SharedViaLink,
      SharedWithMe,
      SharedWithOthers
    },
    Spaces: {
      DriveResolver: SpaceDriveResolver,
      Projects: SpaceProjects
    },
    Trash: {
      Overview: TrashOverview
    }
  }),
  navItems,
  quickActions,
  translations,
  ready({ router, store, globalProperties }: AppReadyHookArgs) {
    const { $clientService } = globalProperties
    Registry.sdkSearch = new SDKSearch(store, router, $clientService)

    // when discussing the boot process of applications we need to implement a
    // registry that does not rely on call order, aka first register "on" and only after emit.
    eventBus.publish('app.search.register.provider', Registry.sdkSearch)
    globalProperties.$archiverService = new ArchiverService(
      $clientService,
      store.getters.configuration.server || window.location.origin,
      get(store, 'getters.capabilities.files.archivers', [
        {
          enabled: true,
          version: '1.0.0',
          formats: ['tar', 'zip'],
          archiver_url: `${store.getters.configuration.server}index.php/apps/files/ajax/download.php`
        }
      ])
    )
  }
}
