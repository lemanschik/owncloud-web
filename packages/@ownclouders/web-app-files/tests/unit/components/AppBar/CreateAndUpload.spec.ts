import CreateAndUpload from 'web-app-files/src/components/AppBar/CreateAndUpload.vue'
import { mock, mockDeep } from 'jest-mock-extended'
import { Resource, SpaceResource } from 'web-client/src/helpers'
import { UppyResource } from 'web-runtime/src/composables/upload'
import { Drive } from 'web-client/src/generated'
import { eventBus, useRequest } from 'web-pkg'
import {
  createStore,
  defaultPlugins,
  shallowMount,
  defaultStoreMockOptions,
  defaultComponentMocks
} from 'web-test-helpers'
import { RouteLocation } from 'vue-router'

jest.mock('web-pkg/src/composables/authContext')

const elSelector = {
  component: '#create-and-upload-actions',
  newFileButton: '#new-file-menu-btn',
  uploadBtn: '#upload-menu-btn',
  resourceUpload: 'resource-upload-stub',
  newFolderBtn: '#new-folder-btn',
  clipboardBtns: '#clipboard-btns',
  pasteFilesBtn: '.paste-files-btn',
  clearClipboardBtn: '.clear-clipboard-btn'
}

const fileHandlerMocks = [
  {
    ext: 'txt',
    action: {
      app: 'text-editor',
      extension: 'txt'
    },
    menuTitle: () => 'Plain text file'
  },
  {
    ext: 'md',
    action: {
      app: 'text-editor',
      extension: 'md'
    },
    menuTitle: () => 'Mark-down file'
  },
  {
    ext: 'drawio',
    action: {
      app: 'draw-io',
      routeName: 'draw-io-edit',
      extension: 'drawio'
    },
    menuTitle: () => 'Draw.io document'
  }
]

describe('CreateAndUpload component', () => {
  describe('action buttons', () => {
    it('should show and be enabled if file creation is possible', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.findComponent<any>(elSelector.uploadBtn).props().disabled).toBeFalsy()
      expect(wrapper.findComponent<any>(elSelector.newFolderBtn).props().disabled).toBeFalsy()
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('should be disabled if file creation is not possible', () => {
      const currentFolder = mockDeep<Resource>({ canUpload: () => false })
      const { wrapper } = getWrapper({ currentFolder })
      expect(wrapper.findComponent<any>(elSelector.uploadBtn).props().disabled).toBeTruthy()
      expect(wrapper.findComponent<any>(elSelector.newFolderBtn).props().disabled).toBeTruthy()
    })
    it('should not be visible if file creation is not possible on a public page', () => {
      const currentFolder = mockDeep<Resource>({ canUpload: () => false })
      const { wrapper } = getWrapper({ currentFolder, currentRouteName: 'files-public-link' })
      expect(wrapper.find(elSelector.component).exists()).toBeFalsy()
    })
  })
  describe('file handlers', () => {
    it('should always show for uploading files and folders', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.findAll(elSelector.resourceUpload).length).toBe(2)
    })
    it('should show additional handlers', () => {
      const { wrapper } = getWrapper({ newFileHandlers: fileHandlerMocks })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('clipboard buttons', () => {
    it('should show if clipboard is empty', () => {
      const { wrapper } = getWrapper()
      expect(
        wrapper.findComponent<any>(`${elSelector.clipboardBtns} oc-button-stub`).exists()
      ).toBeFalsy()
    })
    it('should show if clipboard is not empty', () => {
      const { wrapper } = getWrapper({ clipboardResources: [mockDeep<Resource>()] })
      expect(wrapper.findAll(`${elSelector.clipboardBtns} .oc-button`).length).toBe(2)
    })
    it('call the "paste files"-action', async () => {
      const { wrapper, storeOptions } = getWrapper({ clipboardResources: [mockDeep<Resource>()] })
      await wrapper.find(elSelector.pasteFilesBtn).trigger('click')
      expect(storeOptions.modules.Files.actions.pasteSelectedFiles).toHaveBeenCalled()
    })
    it('call "clear clipboard"-action', async () => {
      const { wrapper, storeOptions } = getWrapper({ clipboardResources: [mockDeep<Resource>()] })
      await wrapper.find(elSelector.clearClipboardBtn).trigger('click')
      expect(storeOptions.modules.Files.actions.clearClipboardFiles).toHaveBeenCalled()
    })
  })
  describe('method "onUploadComplete"', () => {
    it.each([
      { driveType: 'personal', updated: 1 },
      { driveType: 'project', updated: 1 },
      { driveType: 'share', updated: 0 },
      { driveType: 'public', updated: 0 }
    ])('updates the space quota for supported drive types', async (data) => {
      const { driveType, updated } = data
      const { wrapper, mocks, storeOptions } = getWrapper()
      const file = mockDeep<UppyResource>({ meta: { driveType } })
      const graphMock = mocks.$clientService.graphAuthenticated
      graphMock.drives.getDrive.mockResolvedValue(mockDeep<Drive>() as any)
      await wrapper.vm.onUploadComplete({ successful: [file] })
      expect(
        storeOptions.modules.runtime.modules.spaces.mutations.UPDATE_SPACE_FIELD
      ).toHaveBeenCalledTimes(updated)
    })
    it('reloads the file list if files were uploaded to the current path', async () => {
      const eventSpy = jest.spyOn(eventBus, 'publish')
      const itemId = 'itemId'
      const space = mock<SpaceResource>({ id: '1' })
      const { wrapper, mocks } = getWrapper({ itemId, space })
      const file = mockDeep<UppyResource>({
        meta: { driveType: 'project', spaceId: space.id, currentFolderId: itemId }
      })
      const graphMock = mocks.$clientService.graphAuthenticated
      graphMock.drives.getDrive.mockResolvedValue(mockDeep<Drive>() as any)
      await wrapper.vm.onUploadComplete({ successful: [file] })
      expect(eventSpy).toHaveBeenCalled()
    })
  })
  describe('drop target', () => {
    it('is being initialized when user can upload', () => {
      const { mocks } = getWrapper()
      expect(mocks.$uppyService.useDropTarget).toHaveBeenCalled()
    })
    it('is not being initialized when user can not upload', () => {
      const currentFolder = mockDeep<Resource>({ canUpload: () => false })
      const { mocks } = getWrapper({ currentFolder })
      expect(mocks.$uppyService.useDropTarget).not.toHaveBeenCalled()
    })
  })
})

function getWrapper({
  newFileHandlers = [],
  clipboardResources = [],
  files = [],
  currentFolder = mockDeep<Resource>({ canUpload: () => true }),
  currentRouteName = 'files-spaces-generic',
  space = mock<SpaceResource>(),
  item = undefined,
  itemId = undefined,
  newFileAction = false
} = {}) {
  jest.mocked(useRequest).mockImplementation(() => ({
    makeRequest: jest.fn().mockResolvedValue({ status: 200 })
  }))
  const storeOptions = {
    ...defaultStoreMockOptions,
    getters: {
      ...defaultStoreMockOptions.getters,
      newFileHandlers: () => newFileHandlers,
      user: () => ({ id: '1' }),
      capabilities: () => ({
        spaces: { enabled: true },
        files: { app_providers: [{ new_url: '/' }] }
      })
    }
  }
  storeOptions.getters.apps.mockImplementation(() => ({
    fileEditors: []
  }))
  storeOptions.modules.Files.getters.currentFolder.mockImplementation(() => currentFolder)
  storeOptions.modules.Files.getters.clipboardResources.mockImplementation(() => clipboardResources)
  storeOptions.modules.Files.getters.files.mockImplementation(() => files)
  const store = createStore(storeOptions)
  const mocks = {
    ...defaultComponentMocks({ currentRoute: mock<RouteLocation>({ name: currentRouteName }) })
  }
  return {
    storeOptions,
    mocks,
    wrapper: shallowMount(CreateAndUpload as any, {
      data: () => ({ newFileAction }),
      props: { space: space as any, item, itemId },
      global: {
        stubs: { OcButton: false },
        renderStubDefaultSlot: true,
        mocks,
        plugins: [...defaultPlugins(), store]
      }
    })
  }
}
