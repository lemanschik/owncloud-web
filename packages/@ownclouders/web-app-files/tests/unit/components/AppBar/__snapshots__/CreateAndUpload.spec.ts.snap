// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAndUpload component action buttons should show and be enabled if file creation is possible 1`] = `
<div class="create-and-upload-actions oc-flex-inline oc-width-1-1">
  <span>
    <button aria-label="Create a new folder" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-primary oc-button-primary-filled" id="new-folder-btn" type="button">
      <!-- @slot Content of the button -->
      <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="resource-type-folder" size="medium" type="span" variation="passive"></oc-icon-stub>
      <span>New Folder</span>
    </button>
  </span>
  <span>
    <button aria-label="Upload files or folders" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-outline" id="upload-menu-btn" type="button">
      <!-- @slot Content of the button -->
      <oc-icon-stub accessiblelabel="" color="" filltype="line" name="upload" size="medium" type="span" variation="passive"></oc-icon-stub>
      <span>Upload</span>
    </button>
  </span>
  <oc-drop-stub class="oc-width-auto" closeonclick="true" dropid="upload-menu-drop" isnested="false" mode="click" paddingsize="small" position="bottom-start" toggle="#upload-menu-btn">
    <oc-list-stub id="upload-list" raw="false">
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnclass="oc-width-1-1" btnlabel="" isfolder="false"></resource-upload-stub>
      </li>
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnclass="oc-width-1-1" btnlabel="" isfolder="true"></resource-upload-stub>
      </li>
    </oc-list-stub>
  </oc-drop-stub>
  <!--v-if-->
</div>
`;

exports[`CreateAndUpload component file handlers should show additional handlers 1`] = `
<div class="create-and-upload-actions oc-flex-inline oc-width-1-1">
  <span>
    <button aria-label="Create new files or folders" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-primary oc-button-primary-filled" id="new-file-menu-btn" type="button">
      <!-- @slot Content of the button -->
      <oc-icon-stub accessiblelabel="" color="" filltype="fill" name="add" size="medium" type="span" variation="passive"></oc-icon-stub>
      <span>New</span>
    </button>
  </span>
  <oc-drop-stub class="oc-width-auto" closeonclick="true" dropid="new-file-menu-drop" isnested="false" mode="click" paddingsize="small" position="bottom-start" toggle="#new-file-menu-btn">
    <oc-list-stub id="create-list" raw="false">
      <li class="create-list-folder oc-menu-item-hover">
        <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw" id="new-folder-btn" type="button">
          <!-- @slot Content of the button -->
          <oc-resource-icon-stub resource="[object Object]" size="medium"></oc-resource-icon-stub>
          <span>Folder</span>
        </button>
      </li>
      <li class="create-list-file oc-menu-item-hover">
        <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw new-file-btn-txt" type="button">
          <!-- @slot Content of the button -->
          <oc-resource-icon-stub resource="[object Object]" size="medium"></oc-resource-icon-stub>
          <span>Plain text file</span>
        </button>
      </li>
      <li class="create-list-file oc-menu-item-hover">
        <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw new-file-btn-md" type="button">
          <!-- @slot Content of the button -->
          <oc-resource-icon-stub resource="[object Object]" size="medium"></oc-resource-icon-stub>
          <span>Mark-down file</span>
        </button>
      </li>
      <li class="create-list-file oc-menu-item-hover">
        <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw new-file-btn-drawio" type="button">
          <!-- @slot Content of the button -->
          <oc-resource-icon-stub resource="[object Object]" size="medium"></oc-resource-icon-stub>
          <span>Draw.io document</span>
        </button>
      </li>
    </oc-list-stub>
  </oc-drop-stub>
  <span>
    <button aria-label="Upload files or folders" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-outline" id="upload-menu-btn" type="button">
      <!-- @slot Content of the button -->
      <oc-icon-stub accessiblelabel="" color="" filltype="line" name="upload" size="medium" type="span" variation="passive"></oc-icon-stub>
      <span>Upload</span>
    </button>
  </span>
  <oc-drop-stub class="oc-width-auto" closeonclick="true" dropid="upload-menu-drop" isnested="false" mode="click" paddingsize="small" position="bottom-start" toggle="#upload-menu-btn">
    <oc-list-stub id="upload-list" raw="false">
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnclass="oc-width-1-1" btnlabel="" isfolder="false"></resource-upload-stub>
      </li>
      <li class="oc-menu-item-hover">
        <resource-upload-stub btnclass="oc-width-1-1" btnlabel="" isfolder="true"></resource-upload-stub>
      </li>
    </oc-list-stub>
  </oc-drop-stub>
  <!--v-if-->
</div>
`;
